app/models/user.rb

```diff
@@ -9,4 +9,74 @@ class User < ApplicationRecord
   has_many :posts
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships
+
+  def friends?(user)
+    sql = "select f.*
+            from users u1
+            inner join friendships f on u1.id = f.user_1
+            inner join users u2 on u2.id = f.user_2
+            where status = 'Accepted'
+            AND ((u1.id = ? AND u2.id = ?) OR (u1.id = ? AND u2.id = ?))"
+    User.find_by_sql([sql, id, user.id, user.id, id]).first
+  end
+
+  def pending_friend_requests
+    sql = "select u2.*, status
+            from users u1
+            inner join friendships f on u1.id = f.user_1
+            inner join users u2 on u2.id = f.user_2
+            where status = 'Pending'
+            and u1.id = ?"
+    User.find_by_sql [sql, id]
+  end
+
+  def friend_responses
+    sql = "select u1.*, status
+            from users u1
+            inner join friendships f on u1.id = f.user_1
+            inner join users u2 on u2.id = f.user_2
+            where (status = 'Pending' OR status = 'Rejected')
+            and u2.id = ?"
+    User.find_by_sql [sql, id]
+  end
+
+  def pending_friend_responses
+    sql = "select u1.*, status
+            from users u1
+            inner join friendships f on u1.id = f.user_1
+            inner join users u2 on u2.id = f.user_2
+            where status = 'Pending'
+            and u2.id = ?"
+    User.find_by_sql [sql, id]
+  end
+
+  def rejected_friend_responses
+    sql = "select u1.*, status
+            from users u1
+            inner join friendships f on u1.id = f.user_1
+            inner join users u2 on u2.id = f.user_2
+            where status = 'Rejected'
+            and u2.id = ?"
+    User.find_by_sql [sql, id]
+  end
+
+  def relative_posts
+    sql = "select posts.*, u.name
+            from posts
+            inner join users u on posts.user_id = u.id
+            where user_id in
+              (select user_1
+               from friendships
+               where (user_1 = ? or user_2 = ?)
+               and status = 'Accepted'
+               union
+               select user_2
+               from friendships
+               where (user_1 = ? or user_2 = ?)
+               and status = 'Accepted')
+            or user_id = ?
+            order by created_at desc"
+    Post.find_by_sql [sql, id, id, id, id, id]
```

Comment:
I will strongly suggest you use Rails ActiveRecord ORM methods instead of Raw SQL,  it can be easily prone to SQL injection when data are not properly sanitized.

The framework encapsulates it for you and the benefit is that you have to write less code. The other benefit is database independence. The more direct queries you write, the more likely you'll write something that will break when you switch database engines.

See this article: [SQL Injection Prevention Techniques for Ruby on Rails Web Applications](https://www.netsparker.com/blog/web-security/sql-injection-ruby-on-rails-development/)

Reply:
Parameterized statements are not prone to SQL injection.
[Source](https://www.hacksplaining.com/prevention/sql-injection)